File: Tags/DP/673. Number of Longest Increasing Subsequence/673. Number of Longest Increasing Subsequence.js
/**
 * @param {number[]} nums
 * @return {number}
 */
var findNumberOfLIS = function (nums) {
  let dp = new Array(nums.length).fill(1);
  let count = new Array(nums.length).fill(1);
  let max = 0;
  for (let i = 0; i < nums.length; i++) {
    let num = nums[i];
    for (let j = 0; j < i; j++) {
      if (nums[j] < num) {
        if (dp[j] + 1 > dp[i]) {
          dp[i] = dp[j] + 1;
          count[i] = count[j];
        } else if (dp[j] + 1 === dp[i]) {
          count[i] += count[j];
        }
      }
    }
    max = Math.max(max, dp[i]);
  }
  let res = 0;
  for (let i = 0; i < nums.length; i++) {
    if (dp[i] === max) {
      res += count[i];
    }
  }
  return res;
};

File: Day Challenge/20230721 Number of Longest Increasing Subsequence/673. Number of Longest Increasing Subsequence.py
class Solution:
    def findNumberOfLIS(self, nums: List[int]) -> int:
        dp = [1 for _ in range(len(nums))]
        cnt = [1 for _ in range(len(nums))]
        m = 0
        for i in range(0,len(nums)):
            for j in range(0, i):
                if nums[i] > nums[j]:
                    if dp[j] + 1 > dp[i]:
                        dp[i] = dp[j] + 1
                        cnt[i] = cnt[j]
                    elif dp[i] == dp[j] + 1:
                        cnt[i] += cnt[j]
            m = max(m, dp[i])

        ans = 0
        for i in range(len(nums)):
            if dp[i] == m:
                ans += cnt[i]
        return ans